docs/source/llm/run-with-c-plus-plus.md

```diff
@@ -0,0 +1,240 @@
+
+#### `void stop()`
+
+Immediately stops the generation loop. This is typically called from another thread to interrupt a long-running generation.
+
+### GenerationConfig Structure
+
+The `GenerationConfig` struct controls various aspects of the generation process:
+
+```cpp
+struct GenerationConfig {
+  bool echo = true;                // Whether to echo the input prompt in the output
+  int32_t max_new_tokens = -1;     // Maximum number of new tokens to generate
+  bool warming = false;            // Whether this is a warmup run
+  int32_t seq_len = -1;            // Maximum number of total tokens
+  float temperature = 0.8f;        // Temperature for sampling
+  int32_t num_bos = 0;             // Number of BOS tokens to add
+  int32_t num_eos = 0;             // Number of EOS tokens to add
+
+  // Helper method to resolve the actual max_new_tokens based on constraints
+  int32_t resolve_max_new_tokens(int32_t max_context_len, int32_t num_prompt_tokens) const;
+};
+```
+
+The `resolve_max_new_tokens` method handles the complex logic of determining how many tokens can be generated based on:
+- The model's maximum context length
+- The number of tokens in the prompt
+- The user-specified maximum sequence length and maximum new tokens
+
+### Implementation Components
+
+The runner library consists of several specialized components that work together:
+
+#### TextLLMRunner
+
+The main implementation of the `IRunner` interface that orchestrates the text generation process. It manages:
+
+1. Tokenization of input text
+2. Prefilling the KV cache with prompt tokens
+3. Generating new tokens one by one
+4. Collecting performance statistics
+
+#### TextPrefiller
+
+Responsible for processing the initial prompt tokens and filling the KV cache. Key features:
+
+- Efficiently processes large prompts
+- Handles dynamic sequence lengths
+- Supports parallel prefilling for performance optimization
+
+#### TextTokenGenerator
+
+Generates new tokens one by one in an autoregressive manner. It:
+
+- Manages the token generation loop
+- Applies temperature-based sampling
+- Detects end-of-sequence conditions
+- Streams tokens as they're generated
+
+#### TextDecoderRunner
+
+Interfaces with the ExecuTorch Module to run the model forward pass. It:
+
+- Manages inputs and outputs to the model
+- Handles KV cache updates
+- Converts logits to tokens via sampling
+
+## Model Metadata
+
+The metadata includes several important configuration parameters:
+
+1. **`enable_dynamic_shape`**: Whether the model supports dynamic input shapes
+2. **`max_seq_len`**: Maximum sequence length the model can handle
+3. **`max_context_len`**: Maximum context length for KV cache
+4. **`use_kv_cache`**: Whether the model uses KV cache for efficient generation
+5. **`use_sdpa_with_kv_cache`**: Whether the model uses the custom op [`torch.ops.llama.sdpa_with_kv_cache.default`](https://github.com/pytorch/executorch/blob/release/0.7/extension/llm/custom_ops/op_sdpa.cpp#L611-L614)
```

Comment:
Are these in the metadata? Do we and why do we need to specify them for the runner?

Reply:
To control the behaviors of the runner. For this one we may not needed anymore, I can remove it.